cohortextractor/tpp_backend.py

```diff
@@ -2787,6 +2788,207 @@ def patients_outpatient_appointment_date(
+            ]
+            for indication in indication_matches:
+                assert (
+                    indication in valid_indications
+                ), f"'{indication}' is not a valid indication; options are {', '.join(valid_indications)}"
+            indications = [quote(indication) for indication in indication_matches]
+            filter_conditions.append(f"COVID_indication IN ({', '.join(indications)})")
+
+        date_condition, date_joins = self.get_date_condition(
+            "t", "TreatmentStartDate", between
+        )
+        filter_conditions.append(date_condition)
+
+        where_filter_conditions = (
+            f"WHERE {' AND '.join(filter_conditions)}" if filter_conditions else ""
+        )
+
+        # Result ordering
+        if find_first_match_in_period:
+            ordering = "ASC"
+        else:
+            ordering = "DESC"
+
+        use_partition_query = True
+        if returning == "binary_flag" or returning == "date":
+            returning = "binary_flag"
+            column_definition = "1"
+        elif returning == "therapeutic":
+            # remove whitespace, convert to comma-separated string
+            column_definition = "REPLACE(LTRIM(RTRIM(Intervention)), ' and ', ',')"
+        elif returning == "risk_group":
+            # First remove any "Patients with a" and replace " and " with "," within individual risk group fields
+            # Then join the 3 risk cohort fields with ","
+            # Note that the last "s" in SOT02_risk_cohorts is correct
+            parts = [
+                f"REPLACE(REPLACE({risk_col}, 'Patients with a ', ''), ' and ', ',')"
+                for risk_col in [
+                    "MOL1_high_risk_cohort",
+                    "SOT02_risk_cohorts",
+                    "CASIM05_risk_cohort",
+                ]
+            ]
+            # coalesce the parts with a leading ','
+            coalesced_parts = " + ".join(
+                f"coalesce(',' + NULLIF({part}, ''), '')" for part in parts
+            )
+            # use stuff to remove the first ','
+            column_definition = f"STUFF({coalesced_parts}, 1, 1, '')"
+        elif returning == "region":
+            column_definition = "Region"
+        elif returning == "number_of_matches_in_period":
+            column_definition = "COUNT(*)"
+            use_partition_query = False
+        else:
+            raise ValueError(f"Unsupported `returning` value: {returning}")
+
+        if use_partition_query:
+            # There can be duplicates per patient in the Therapeutics dataset
+            # These are likely to be invalid in some way, but we keep them in so users
+            # can identify them and deal with them as appropriate
+            # (We remove fully duplicate rows in the temp table only)
+            # Duplicate rows are sorted by all the fields that have been identified to
+            # contain duplicate values for a patient, to ensure a consistent return value
```

Comment:
I misunderstood this comment at first but I understand now having seen the conversation on the original ticket. So we only ever return one result per patient (as we must) but when we count results we included duplicates and so this is allows researchers to spot that the duplicates are present. Is that right?

Reply:
Yes, so there can be duplicates per patient, which differ on one or more fields, but the assumption (for now at least) is that those duplicates are errors.   Mostly the values are the same apart for one or two fields.   We only ever return one row per patient, either the first or last by TreatmentStartDate, but in the case where there are duplicates, we want the sorting to be consistent.  There's no sensible way to pick which e.g. region is the correct one, if all other fields are the same, so this just makes sure that it always returns the same one (the first, alphabetically).  The idea is that researchers can use the `number_of_matches_in_period` return value to check if they have duplicates and exclude those patients from further analysis.

I'll see if I can make this comment more understandable :)